### DISCUSSION TITLE
CLI Settings Validation

### DISCUSSION THREAD

**Discussion by turp** [spectreconsole/spectre.console#217] [Q&A] [answered]
Does the CLI framework contain a way to validate required settings properties before a command is executed? I have a number of commands with required and optional parameters, but it's not clear how to enforce that behavior. Any guidance would be appreciated.

**Comment by turp** [comment]
I did find that there's a ```Validate()``` method in both the Command and CommandSettings classes and that helped answer part of my question. And when I explored the tests, it looks like you can create Validator attributes, like:

```
[EvenNumberValidator("Animals must have an even number of legs.")]
[PositiveNumberValidator("Number of legs must be greater than 0.")]
```

**Comment by patriksvensson** [comment, marked as answer]
@turp Here is an example of how to validate arguments/options, and how to set arguments as optional/required and option values as optional/required. Reach out if you have any questions. Below the code example, there are some images showing the results of different operations.

# Code

```csharp
using Spectre.Console;
using Spectre.Console.Cli;
using System;
using System.ComponentModel;

namespace ConsoleApp9
{
    public class Program
    {
        public static int Main(string[] args)
        {
            var app = new CommandApp<MyCommand>();
            return app.Run(args);
        }
    }

    public sealed class MyCommand : Command<MyCommandSettings>
    {
        // Validate as part of the command.
        // This is a good way of validating options if you
        // require any injected services.
        public override ValidationResult Validate(CommandContext context, MyCommandSettings settings)
        {
            if (settings.RequiredArgument == "Hello")
            {
                return ValidationResult.Error("Required argument cannot be 'Hello'");
            }

            return ValidationResult.Success();
        }

        public override int Execute(CommandContext context, MyCommandSettings settings)
        {
            Console.WriteLine("RequiredArgument: {0}", settings.RequiredArgument);
            Console.WriteLine("OptionalArgument: {0}", settings.OptionalArgument);
            Console.WriteLine("OptionWithRequiredValue: {0}", settings.OptionWithRequiredValue);
            Console.WriteLine("OptionWithOptionalValue: {0}", settings.OptionWithOptionalValue);

            return 0;
        }
    }

    public sealed class MyCommandSettings : CommandSettings
    {
        [CommandArgument(0, "<FOO>")]
        [Description("An argument that is required")]
        public string RequiredArgument { get; set; }

        [CommandArgument(1, "[BAR]")]
        [Description("An argument that is completely optional.")]
        public string OptionalArgument { get; set; }

        [CommandOption("--required <BAZ>")]
        [Description("An option with a required value. The option is not required per se.")]
        public string OptionWithRequiredValue { get; set; }

        [CommandOption("--optional [QUX]")]
        [Description("An option with an optional value.")]
        public FlagValue<string> OptionWithOptionalValue { get; set; }

        // Validate as part of the setting.
        // This is a good way of validating options if you don't
        // require any injected services.
        public override ValidationResult Validate()
        {
            if (OptionWithRequiredValue == "Hello")
            {
                return ValidationResult.Error(
                    "Required option ([yellow]--required[/]) cannot be 'Hello'");
            }

            return ValidationResult.Success();
        }
    }
}
```

# Examples

![image](https://user-images.githubusercontent.com/357872/104164228-e9d4d880-53f7-11eb-8ebc-f7d18fe73119.png)
![image](https://user-images.githubusercontent.com/357872/104164280-f9ecb800-53f7-11eb-93ab-77f94203781b.png)
![image](https://user-images.githubusercontent.com/357872/104164519-5059f680-53f8-11eb-807c-ea1020c5360d.png)
![image](https://user-images.githubusercontent.com/357872/104164560-5d76e580-53f8-11eb-8422-f563758e991d.png)

**Reply by turp** [reply]
Thanks @patriksvensson - that helps a lot.

**Reply by osmant** [reply]
```c#
            if (OptionWithRequiredValue == "Hello")
            {
                return ValidationResult.Error(
                    "Required option ([yellow]--required[/]) cannot be 'Hello'");
            }`
```

Is the code I copied from your example supposed to print `--required` word with yellow color ? If so, I am unable to get it work like that. Please see the image below how it looks when the word is printed. Is this an expected behavior ?

![image](https://user-images.githubusercontent.com/3219424/126856437-70556ae1-9210-4185-adc1-e1ecd37dc005.png)

**Reply by turp** [reply]
You're correct @osmant - i get the same results. I don't think you can send markup in the ValidationResult message - it looks like it is stripped out when processing the exception in CommandApp line 136 https://github.com/spectreconsole/spectre.console/blob/57731c0d550162b9eddc60f16c9e2698559cf0b6/src/Spectre.Console/Cli/CommandApp.cs#L136

**Reply by patriksvensson** [reply]
Yes, that is correct. It was a mistake in the example.
